source/mesh_refinement/thermal_energy_density.cc

```diff
@@ -101,11 +101,36 @@ namespace aspect
                                       this->get_mpi_communicator());
       vec = vec_distributed;
 
-      DerivativeApproximation::approximate_gradient  (this->get_mapping(),
-                                                      this->get_dof_handler(),
-                                                      vec,
-                                                      indicators,
-                                                      this->introspection().component_indices.temperature);
+      try
+        {
+          DerivativeApproximation::approximate_gradient  (this->get_mapping(),
+                                                          this->get_dof_handler(),
+                                                          vec,
+                                                          indicators,
+                                                          this->introspection().component_indices.temperature);
+        }
+      catch (std::exception &exc)
+        {
+          std::cerr << std::endl << std::endl
+                    << "----------------------------------------------------"
+                    << std::endl;
+          std::cerr << "Exception on MPI process <"
+                    << Utilities::MPI::this_mpi_process(MPI_COMM_WORLD)
+                    << "> while running mesh refinement plugin <"
+                    << "thermal energy density"
+                    << ">: " << std::endl
+                    << "Mesh refinement plugin thermal energy density could failed to find a gradient. "
```

Comment:
I would reword this a bit, e.g. like this:
```suggestion
                    << "The mesh refinement plugin failed to compute a thermal energy density gradient. "
```